pkg/emb6/contrib/sock/udp/emb6_sock_udp.c

```diff
@@ -0,0 +1,317 @@
+    assert((remote == NULL) || (remote->port != 0));
+    if (sock->sock.input_callback != NULL) {
+        sock_udp_close(sock);
+    }
+    mutex_init(&sock->mutex);
+    mutex_lock(&sock->mutex);
+    mbox_init(&sock->mbox, sock->mbox_queue, SOCK_MBOX_SIZE);
+    atomic_flag_clear(&sock->receivers);
+    if ((res = _reg(&sock->sock, sock, _input_callback, local, remote)) < 0) {
+        sock->sock.input_callback = NULL;
+        mutex_unlock(&sock->mutex);
+        return res;
+    }
+    mutex_unlock(&sock->mutex);
+    return 0;
+}
+
+void sock_udp_close(sock_udp_t *sock)
+{
+    assert(sock != NULL);
+    if (sock->sock.input_callback != NULL) {
+        while (atomic_fetch_sub(&sock->receivers, 1) > 0) {
+            msg_t msg = { .type = _MSG_TYPE_CLOSE };
+            mbox_put(&sock->mbox, &msg);
+        }
+        mutex_lock(&sock->mutex);
+        udp_socket_close(&sock->sock);
+        sock->sock.input_callback = NULL;
+        mutex_unlock(&sock->mutex);
+    }
+}
+
+int sock_udp_get_local(sock_udp_t *sock, sock_udp_ep_t *ep)
+{
+    assert((sock != NULL) && (ep != NULL));
+    if ((sock->sock.input_callback != NULL) &&
+        (sock->sock.udp_conn->lport != 0)) {
+        mutex_lock(&sock->mutex);
+        memset(&ep->addr, 0, sizeof(ipv6_addr_t));
+        ep->port = NTOHS(sock->sock.udp_conn->lport);
+        mutex_unlock(&sock->mutex);
+        return sizeof(ipv6_addr_t);
+    }
+    return -EADDRNOTAVAIL;
+}
+
+int sock_udp_get_remote(sock_udp_t *sock, sock_udp_ep_t *ep)
+{
+    assert((sock != NULL) && (ep != NULL));
+    if ((sock->sock.input_callback != NULL) &&
+        (sock->sock.udp_conn->rport != 0)) {
+        mutex_lock(&sock->mutex);
+        memcpy(&ep->addr, &sock->sock.udp_conn->ripaddr, sizeof(ipv6_addr_t));
+        ep->port = NTOHS(sock->sock.udp_conn->rport);
+        mutex_unlock(&sock->mutex);
+        return sizeof(ipv6_addr_t);
+    }
+    return -ENOTCONN;
+}
+
+int sock_udp_recv(sock_udp_t *sock, void *data, size_t max_len,
+                  uint32_t timeout, sock_udp_ep_t *remote)
+{
+    xtimer_t timeout_timer;
+    int blocking = BLOCKING;
+    int res = -EIO;
+    msg_t msg;
+
+    assert((sock != NULL) && (data != NULL) && (max_len > 0));
+    if (sock->sock.input_callback == NULL) {
+        return -EADDRNOTAVAIL;
+    }
+    if (timeout == 0) {
+        blocking = NON_BLOCKING;
+    }
+    else if (timeout != SOCK_NO_TIMEOUT) {
+        timeout_timer.callback = _timeout_callback;
+        timeout_timer.arg = &sock->mbox;
+        xtimer_set(&timeout_timer, timeout);
+    }
+    atomic_fetch_add(&sock->receivers, 1);
+    if (_mbox_get(&sock->mbox, &msg, blocking) == 0) {
+        return -EAGAIN;
+    }
+    switch (msg.type) {
+        case _MSG_TYPE_CLOSE:
+            res = -EADDRNOTAVAIL;
+            break;
+        case _MSG_TYPE_TIMEOUT:
+            res = -ETIMEDOUT;
+            break;
+        case _MSG_TYPE_RCV:
+            mutex_lock(&sock->mutex);
```

Comment:
it is necessary to lock here already, if possible (and safe) move lock below one line above `memcpy`?